src/regmod/composite_models/node.py

```diff
@@ -0,0 +1,57 @@
+"""
+Node Model
+"""
+from typing import Dict
+
+from pandas import DataFrame
+
+
+class NodeModel:
+    """
+    Abstract class that encode the behavior of the node
+    """
+
+    def __init__(self, name: str):
+        self.name = name
+
+    def set_data(self,
+                 df: DataFrame,
+                 col_value: str = None,
+                 col_label: str = None):
+        raise NotImplementedError
+
+    def get_data(self, col_label: str = None) -> DataFrame:
+        raise NotImplementedError
+
+    def fit(self, **fit_options):
+        raise NotImplementedError
+
+    def predict(self,
+                df: DataFrame = None,
+                col_value: str = None,
+                col_label: str = None) -> DataFrame:
+        raise NotImplementedError
+
+    def set_prior(self, priors: Dict, masks: Dict = None):
+        raise NotImplementedError
+
+    def get_posterior(self) -> Dict:
+        raise NotImplementedError
+
+    def subset_df(self,
+                  df: DataFrame,
+                  col_label: str = None,
+                  copy: bool = False) -> DataFrame:
+        if col_label is not None and col_label in df.columns:
+            df = df[df[col_label] == self.name]
+        if copy:
+            df = df.copy()
```

Comment:
A little confused here. Why assign copy of df to df?

Reply:
Just to use the copy of the data instead of the original data, since we return `df` at the end.